README.md
# Top.gg Node SDK

An official module for interacting with the Top.<span>gg API

# Installation

`yarn add @top-gg/sdk` or `npm i @top-gg/sdk`

# Introduction

The base client is Topgg.Api, and it takes your Top.<span>gg token and provides you with plenty of methods to interact with the API.

Your Top.<span>gg token can be found at `top.gg/bot/(BOT_ID)/webhooks` and copying the token.

You can also setup webhooks via Topgg.Webhook, look down below at the examples for how to do so!

# Links

[Documentation](https://topggjs.rtfd.io)

[API Reference](https://docs.top.gg) | [GitHub](https://github.com/top-gg/node-sdk) | [NPM](https://npmjs.com/package/@top-gg/sdk) | [Discord Server](https://discord.gg/EYHTgJX)

# Auto-Posting

If you're looking for an easy way to post your bot's stats (server count, shard count), check out [`topgg-autoposter`](https://npmjs.com/package/topgg-autoposter)

# Popular Examples

## Auto-Posting stats

If you choose not to use [`topgg-autoposter`](https://npmjs.com/package/topgg-autoposter) you can always do it yourself manually;

```js
const client = Discord.Client() // Your discord.js client

const Topgg = require('@top-gg/sdk')

const api = new Topgg.Api('Your top.gg token')

setInterval(() => {
  api.postStats({
    serverCount: client.guilds.cache.size,
    shardId: client.shard.ids[0], // if you're sharding
    shardCount: client.options.shardCount
  })
}, 1800000) // post every 30 minutes
```
With this your server count and shard count will be posted to top.<span>gg

## Webhook server

```js
const express = require('express')
const Topgg = require('@top-gg/sdk')

const app = express() // Your express app

const webhook = new Topgg.Webhook('topggauth123') // add your top.gg webhook authorization (not bot token)

app.post('/dblwebhook', webhook.listener(vote => {
  // vote is your vote object
  console.log(vote.user) // 221221226561929217
})) // attach the middleware

app.listen(3000) // your port
```
With this example, your webhook dashboard should look like this:
![](https://i.imgur.com/wFlp4Hg.png)


docs/Changelog.md
<a name="Change Log"></a>

# Change Log

## 3.1.0

  - Added Webhook.listener() to replace Webhook.middleware()

## 3.0.9
  - A change to how typings are used, properly exports via dist/typings

## 3.0.8
  - Change docs URL
  - Fix typos in a few method docs
  - Specified that .hasVoted() is last 12 hours
  - Auto-respond with status 200 as per docs

## 3.0.7
  - Add new BotInfo.bannerUrl to typings

## 3.0.6
  - Fix to webhook, where there were issues when using a body parser that also used `raw-body`

## 3.0.5
  - Fix to Api.isWeekend example

## 3.0.4
  - Released official documentation page
  - Added GitHub actions to publish to npm on release

## 3.0.3
  - Fixed some webhook bugs
  - Improved a few typings

## 3.0.2
Some day-one patches to README, 3.0.1 is a dud release

## 3.0.0
A full rewrite of the package, here's some of the major changes:
  - The webhook and API interaction have been full separated, into two different classes, `Api` and `Webhook`
  - Auto-Posting is no longer natively supported in the package, instead you can use the officially *supported* [`topgg-autoposter`](https://npmjs.com/topgg-autoposter) package.
  - The webhook class is no longer ran off of events, instead acting as a middleware function that defines `req.vote`


src/structs/API.ts
import fetch, { Headers } from 'node-fetch'
import APIError from '../utils/APIError'
import ClientUtils, { ClientType } from '../utils/ClientUtils'
import qs from 'querystring'
import '../typings'
import { EventEmitter } from 'events'

interface APIOptions {
  /**
   * Top.GG Token
   */
  token: any
  /**
   * Interval at which to post in milliseconds
   * @default 1800000 30 minutes
   */
  interval?: number
  /**
   * Whether or not to autopost
   * @default true
   */
  autoPost?: boolean
  /**
   * Whether or not in testing phase (devs only)
   */
  debug?: boolean
}

/**
 * Top.GG API Client for Posting stats or Fetching data
 * @link https://top.gg/api/docs
 * @example
 * ```js
 * const client = Discord.Client() // Your discord.js or Eris client
 *
 * const TopGG = require('@top-gg/sdk')
 *
 * const api = new TopGG.API('Your top.gg token', client)
 *
 * api.on('posted', () => {
 *   console.log('Server count posted!')
 * })
 * ```
 */
export default class API extends EventEmitter {
  private options: APIOptions
  private client: any
  private clientType: ClientType
  /**
   * Create Top.GG API instance
   * @param token Token or options
   * @param client Supported library client 
   */
  constructor (token?: string|APIOptions, client?: any) {
    super()
    const defaultOptions = {
      interval: 1800000,
      autoPost: true,
      debug: false
    }
    if (!token || typeof token === 'string') {
      this.options = {
        token,
        ...defaultOptions
      }
    } else {
      this.options = {
        ...defaultOptions,
        ...token
      }
    }

    if (this.options.interval < 900000) throw new Error('Interval can not be lower than 900000 (15 minutes)')

    this.client = client 
    this.clientType = ClientUtils.figureClient(client)

    if (this.options.debug) console.debug(`[DEBUG] Client Type: ${this.clientType}`)

    if (this.options.autoPost && client) ClientUtils.setupAutoPoster(client, this.clientType, () => {
      this.postStats()

      setInterval(() => {
        if (ClientUtils.isStarted(client, this.clientType)) this.postStats()
      }, this.options.interval)
    })
  }

  private async _request (method: string, path: string, body?: object): Promise<any> {
    const headers = new Headers()
    if (this.options.token) headers.set('Authorization', this.options.token)
    if (method !== 'GET') headers.set('Content-Type', 'application/json')

    let url = `https://top.gg/api/${path}`

    // @ts-ignore querystring typings are messed
    if (body && method === 'GET') url += `?${qs.stringify(body)}`

    if (this.options.debug) {
      console.debug(`[DEBUG] ${method} ${path}: ${JSON.stringify(body)}`)

      return {
        test: true
      }
    }

    const response = await fetch(url, {
      method,
      headers,
      body: body && method !== 'GET' ? JSON.stringify(body) : null
    })

    const responseBody = response.headers.get('Content-Type')?.startsWith('application/json') ? await response.json() : await response.text()

    if (!response.ok) throw new APIError(response.status, response.statusText, responseBody)

    return responseBody
  }

  /**
   * Post bot stats to Top.GG (Do not use if you supplied a client)
   * @param serverCount Server count
   * @param shardId Current shard ID
   * @param shardCount Total shard count
   * @example
   * ```js
   * await client.postStats(28199, null, 1)
   * ```
   */
  public async postStats (serverCount?: number, shardId?: number, shardCount?: number) {
    if (!serverCount && !this.clientType) throw new Error('Missing Server Count or Client')
    const data = {
      server_count: serverCount ?? ClientUtils.serverCount(this.client, this.clientType),
      shard_id: shardId ?? ClientUtils.currentShard(this.client, this.clientType),
      shard_count: shardCount ?? ClientUtils.shardCount(this.client, this.clientType)
    }

    /**
     * Event when the bot posts it's stats
     * @event posted
     * @param serverCount Posted server count
     * @param shardId Posted shard ID
     * @param shardCount Posted shard count
     */
    this._request('POST', '/bots/stats', data)
      .then(() => this.emit('posted', data.server_count, data.shard_id, data.shard_count))
      .catch(err => console.error(err))
  } 

  /**
   * Get a bots stats
   * @param id Bot ID
   * @example
   * ```js
   * await client.getStats() // gives self
   * await client.getStats('461521980492087297')
   * // =>
   * {
   *   server_count: 28199,
   *   shard_count 1,
   *   shards: []
   * }
   * ```
   */
  public async getStats (id: Snowflake): Promise<BotStats> {
    if (!id && !this.clientType) throw new Error('ID or Client missing')
    return this._request('GET', `/bots/${id ?? ClientUtils.id(this.client, this.clientType)}/stats`)
  }

  /**
   * Get bot info
   * @param id Bot ID
   * @example
   * ```js
   * await client.getBot() // returns self
   * await client.getBot('461521980492087297') // returns other bot
   * ```
   */
  public async getBot (id: Snowflake): Promise<BotInfo> {
    if (!id && !this.clientType) throw new Error('ID Missing')
    return this._request('GET', `/bots/${id ?? ClientUtils.id(this.client, this.clientType)}`)
  }

  /**
   * Get user info
   * @param id User ID
   * @example
   * ```js
   * await client.getUser('205680187394752512')
   * // =>
   * user.username // Xignotic
   * ```
   */
  public async getUser (id: Snowflake): Promise<UserInfo> {
    if (!id) throw new Error('ID Missing')
    return this._request('GET', `/users/${id}`)
  }

  /**
   * Get a list of bots
   * @param query Bot Query
   * @example
   * ```js
   * // Finding by properties
   * await client.getBots({
   *   search: {
   *     username: 'shiro',
   *     certifiedBot: true
   *     ...any other bot object properties
   *   }
   * })
   * // =>
   * {
   *   results: [
   *     {
   *       id: '461521980492087297',
   *       username: 'Shiro',
   *       discriminator: '8764',
   *       lib: 'discord.js',
   *       ...rest of bot object
   *     }
   *     ...other shiro knockoffs B)
   *   ],
   *   limit: 10,
   *   offset: 0,
   *   count: 1,
   *   total: 1
   * }
   * // Restricting fields
   * await client.getBots({
   *   fields: ['id', 'username']
   * })
   * // =>
   * {
   *   results: [
   *     {
   *       id: '461521980492087297',
   *       username: 'Shiro'
   *     },
   *     {
   *       id: '493716749342998541',
   *       username: 'Mimu'
   *     },
   *     ...
   *   ],
   *   ...
   * }
   * ```
   */
  public async getBots (query?: BotsQuery): Promise<BotsResponse> {
    if (query) {
      if (query.fields instanceof Array) query.fields = query.fields.join(', ')
      if (query.search instanceof Object) {
        query.search = Object.keys(query.search)
          .map(key => `${key}: ${query.search[key]}`).join(' ')
      }
    }
    return this._request('GET', '/bots', query)
  }

  /**
   * Get users who've voted
   * @example
   * ```js
   * await client.getVotes()
   * // => 
   * [
   *   {
   *     username: 'Xignotic',
   *     discriminator: '0001',
   *     id: '205680187394752512',
   *     avatar: '3b9335670c7213b3a2d4e990081900c7'
   *   },
   *   {
   *     username: 'iara',
   *     discriminator: '0001',
   *     id: '395526710101278721',
   *     avatar: '3d1477390b8d7c3cec717ac5c778f5f4'
   *   }
   *   ...more
   * ]
   * ```
   */
  public async getVotes (): Promise<Array<ShortUser>> {
    if (!this.options.token) throw new Error('Missing token')
    return this._request('GET', '/bots/votes')
  }

  /**
   * Get whether or not a user has voted
   * @param id User ID
   * @example
   * ```js
   * await client.hasVoted('205680187394752512')
   * // => true/false
   * ```
   */
  public async hasVoted(id: Snowflake): Promise<boolean> {
    if (!id) throw new Error('Missing ID')
    return this._request('GET', '/bots/check', { userId: id }).then(x => !!x.voted)
  }

  /**
   * Whether or not the weekend multiplier is active
   * @example
   * ```js
   * await client.hasVoted()
   * // => true/false
   * ```
   */
  public async isWeekend (): Promise<boolean> {
    return this._request('GET', '/weekend').then(x => x.is_weekend)
  }
}